# Introduction to Machine Learning

Python is probably the most often used programming language to train and run machine learning models.

![sklearn](../static/sklearn.png)

For shallow machine learning models, [`sklearn`](https://scikit-learn.org) is undoubtedly the most popular library you can use. Sklearn has tutorials (check this great [*getting started*](https://scikit-learn.org/stable/getting_started.html) and this [*common pitfalls*](https://scikit-learn.org/stable/common_pitfalls.html) tutorial!), toy-datasets, extensive descriptions for all models, algorithms and many more things. Its incredible and I highly highly recommed you to use this library, if you choose to continue working with machine learning.

When it comes to deep learning, python also offers a lot of options. Personally, I prefer to use [`PyTorch`](https://pytorch.org/), but there are other options. 

Anyway, we wont bother with deep learning in this course, so lets just skip straight to sklearn. As opposed to numpy, I prefer to directly import the classes and methods I use, instead of importing the whole module. I dont know why, but its what people just do, apparently.

The first thing we will do now is fetch the breast cancer dataset from sklearn. It is a common, easy to use dataset that allows us to apply a easy binary classification. It contains features computed from digitized images of a breast mass and a binary label: malignant (0) or benign (1).

In [ ]:
from sklearn.datasets import load_breast_cancer

#import for plotting
import seaborn as sns
import matplotlib.pyplot as plt

We set `as_frame=True` to get both X and y as dataframe objects. Sklearn estimators typically require numpy arrays as their input, but pandas is a nice way to look at your data in a meaningful way

In [20]:

X,y = load_breast_cancer(return_X_y=True,as_frame=True)
print(X.shape,y.shape)

(569, 30) (569,)


To make our data more intuitive, we can create a new variable called `diagnosis` which holds the actual labels that correspond to 0 and 1 in our y variable. We should not use this to train our model, but it makes sense to use this variable for data exploration.

In [84]:
diagnosis = y.map({0:"Malignant",1:"Benignant"})

In [19]:
X.head(n=10)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   
5        12.45         15.70           82.57      477.1          0.12780   
6        18.25         19.98          119.60     1040.0          0.09463   
7        13.71         20.83           90.20      577.9          0.11890   
8        13.00         21.82           87.50      519.8          0.12730   
9        12.46         24.04           83.97      475.9          0.11860   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760         0.30010              0.14710         0.2419   
1           0.07864         0.08690              0.07017         0.1812   
2           0.15990         0.19740              0.12790         0.2069   
3           0.28390         0.24140              0.10520         0.2597   
4           0.13280         0.19800              0.10430         0.1809   
5           0.17000         0.15780              0.08089         0.2087   
6           0.10900         0.11270              0.07400         0.1794   
7           0.16450         0.09366              0.05985         0.2196   
8           0.19320         0.18590              0.09353         0.2350   
9           0.23960         0.22730              0.08543         0.2030   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   
5                 0.07613  ...         15.47          23.75           103.40   
6                 0.05742  ...         22.88          27.66           153.20   
7                 0.07451  ...         17.06          28.14           110.60   
8                 0.07389  ...         15.49          30.73           106.20   
9                 0.08243  ...         15.09          40.68            97.65   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   
5       741.6            0.1791             0.5249           0.5355   
6      1606.0            0.1442             0.2576           0.3784   
7       897.0            0.1654             0.3682           0.2678   
8       739.3            0.1703             0.5401           0.5390   
9       711.4            0.1853             1.0580           1.1050   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  
5                0.1741          0.3985                  0.12440  
6                0.1932          0.3063                  0.08368  
7                0.1556          0.3196                  0.11510  
8                0.2060          0.4378                  0.10720  
9      

In [28]:
y.head(n=10)

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: target, dtype: int32

Inspecting the `shape` of our feature matrix **X**, we see that we have 569 `samples` and 30 `features`.

The ground truth label vector **y** is one-dimensional.

In [30]:
X.shape,y.shape

((569, 30), (569,))

### **Exercise**

It makes a lot of sense to exermine the data you are dealing with. Using the seaborn library, examine the distribution (how many entries) of the `y` variable.

**Hint**: 

You want to use  `sns.histplot` here.

### **Exercise**

Use the `sns.boxplot()` function to plot a distribution plot of a feature of your choice (e.g., a column from **X**), given the labels stored in `diagnosis`.

**Hint**:

Pass diagnosis to the `x` argument and the column of your choice to the `y` argument. Dont forget to be very explicit when passing the arguments to seaborn.

### **Exercise**

Next thing to check would be our input data. We have 32 features ready to use in our dataframe. It makes a lot of sense to check the correlation between these features. If features are highly redundant, they may lead to a overfitting model, because there is a lot of redundancy and noise in the data. Meaning, the model learns features that may not actually be important.

Typically, we want to train (adapt weights!) our model with features that are highly informative.

Since our feature matrix **X** is a `pandas dataframe` we can get the correlation scores between features by calling `df.corr()`. Be sure to replace `df` with your own dataframe (X),

Create a variable called `corr` and store in that variable the correlation scores. Then, use `sns.heatmap()` to plot the correlation scores. Do you notice any interesting trends? 

**Hint**:

Try to pass the `cmap="coolwarm"` argument to the `sns.heatmap` function. 

You can exclude features before you actually train your model, especially if you suspect that they might influence the performance of your model. Since we only have 32 features, we can still go on with this dataset.

One very very important thing you want to do before you train your model is to `normalize` your features! Non-normalized features can heavily influence the performance of your model. You dont want that to happen to you.

Using the `df.describe()` method, we see that all features differ heavily in their mean, min, max and std value.

In [54]:
X.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

Luckily, sklearn has a solution for this. We can use the `StandardScaler` function from the `preprocessing` module. `StandardScaler` allows us to scale our data to a mean of 0 and unit variance.

          z = (x - u) / s

In [55]:
from sklearn.preprocessing import StandardScaler

But now one very important thing before we proceed!

Scaling your data is super important, **but** we only want to scale our `training data`! If we scale the whole dataset, we will contamine our test data, by leaking distribution information from the training set to the test set.

This will render all our conclusions and our analysis meaningless.

Thats why we first want to split our dataset into a `training` and `test` set.

Of course, sklearn offers us a easy solution, using the `train_test_split` function.

In [72]:
from sklearn.model_selection import train_test_split

There are some key arguments we need to pass here. First, we need to pass our full dataset, so X and y.

Then, we need to define the size of the split using the `train_size` argument. The size of this argument depends a lot on the data size you have available, choosing 70% to 80% is a good rule of thumb here. 

At last, we can also set the `random_state` argument to enhance reproducibility.

In [75]:
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.8,random_state=42)

### **Exercise** 

Inspect the shape of your training and test data to see, if your split worked as intended. You want to see less samples in your test data and more samples in your training data.

Now with this out of the way, we can now actually scale our data. We want to call the `Scaler.fit_transform` on our training data and `Scaler.transform` on our test data.

This way, we make sure, that the test data is scaled using the training distribution, and there is no leakage beforehand.

In [ ]:
Scaler = StandardScaler()

X_train_scaled = Scaler.fit_transform(X_train)
X_test_scaled = Scaler.transform(X_test)

### **Exercise**

Using the np.mean() function, examine how the scaled and the non scaled arrays differ. 

**Hint**: You can use the `axis` argument, to either view the mean of the samples (rows) or the features (columns). axis = 0 will give you the mean per column (average column across all samples)